charmhelpers/contrib/sysctl/watermark_scale_factor.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright 2014-2015 Canonical Limited.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from charmhelpers.core.hookenv import (
+    log,
+    DEBUG,
+    ERROR,
+)
+
+import re
+
+WMARK_MAX = 1000
+WMARK_DEFAULT = 10
+MEM_TOTAL_MIN_KB = 16777152
+MAX_PAGES = 2500000000
+P = re.compile('\d+')
+
+
+def calculate_watermark_scale_factor():
+    """Calculates optimal vm.watermark_scale_factor value
+
+    :returns: watermark_scale_factor
+    :type: int
+    """
+
+    mem_total = get_memtotal()
+    normal_managed_pages = get_normal_managed_pages()
+
+    try:
+        wmark = min([watermark_scale_factor(mem_total, managed_pages)
+                     for managed_pages in normal_managed_pages])
+    except ValueError as e:
+        log("Failed to calculate watermark_scale_factor from normal managed pages: {}".format(normal_managed_pages), ERROR)
+        raise e
+
+    log("vm.watermark_scale_factor: {}".format(wmark), DEBUG)
+    return wmark
+
+
+def get_memtotal():
+    memtotal = None
+    try:
+        with open('/proc/meminfo', 'r') as f:
+            for line in f:
+                print(line)
+                if "MemTotal" in line:
+                    memtotal = int(P.search(line).group())
+                    break
+                else:
+                    raise Exception("Could not find MemTotal")
+    except (Exception, OSError) as e:
+        log("Failed to parse /proc/meminfo in calculating watermark_scale_factor: {}".format(e), ERROR)
+        raise e
+
+    return memtotal
+
+
+def get_normal_managed_pages():
+    try:
+        normal_managed_pages = []
+        with open('/proc/zoneinfo', 'r') as f:
+            for line in f:
+                if "Node" in line and "zone" in line:
+                    zone = [v for v in line.split(' ')
+                            if v in ["DMA", "DMA32", "Normal", "Movable", "Device"]][0]
+
+                if zone == "Normal" and "managed" in line:
+                    managed = int([v for v in line.split(' ') if P.match(v)][0])
+                    normal_managed_pages.append(managed)
```

Comment:
So a few issues with the code here:

1. `zone` is potentially accessed at line 81 before it is assigned at 78 (due to the if statement).  Before the loop, a
  `zone = None` ought to be done to eliminate a variable not found type error.
2. The code is using the string in `zone` as a boolean to indicate whether it is parsing a "Node <d>, zone Normal" section in the `/proc/zoneinfo` output.  Can we make this more obvious by:
  * A comment on what the block of code is doing.
  * Using a boolean to indicate that it is in the 'Normal Zone'

I might do something like the following:

```python
with open(...) as f:
    in_zone_normal = False
    # regex to search for strings that look like "Node 0, zone    Normal" and last string to group 1
    normal_zone_matcher = re.compiile(r"^Node\s\d+\s+zone\s+(\S+)$")
    # regex to match to a number at the end of the line.
    managed_matcher = re.compile(r"\s+managed\s+(\d+)$")
    for line in f:
        # See if the line is at the beginning of a Node section, capturing it to group 1
        match = normal_zone_matcher.search(line)
        if match:
            in_zone_normal = match.group(1) == 'Normal'
        if in_zone_normal:
            # match the number at the end of "     managed    3840" into group 1.
             managed_match = managed_matcher.search(line)
             if managed_match:
                normal_managed_pages.append(managed_match.group(1))
                # skip the rest of the section
                in_zone_normal = False
```

Note, I'm only using a regex to make it clear that we are searching for a particular type of line to then start sub-searching for the 'managed' part in the line.